corehq/tests/locks.py

```diff
@@ -0,0 +1,80 @@
+import logging
+from contextlib import contextmanager
+from threading import RLock
+from unittest.mock import patch
+
+import attr
+
+import dimagi.utils.couch as module
+
+log = logging.getLogger(__name__)
+
+
+def reentrant_redis_locks(test=None):
+    """Decorator/context manager to enable reentrant redis locks
+
+    This is useful for tests that do things like acquire a lock and then
+    fire off a celery task (which will usually be executed synchronously
+    in tests) before the lock is released.
+
+    Usage as decorator:
+
+        @reentrant_redis_locks
+        def test_something():
+            ...
+
+    Usage as context manager:
+
+        def test_something():
+            with reentrant_redis_locks():
+                ...
+    """
+    def get_reentrant_lock(key, **kw):
+        try:
+            lock = locks[key]
```

Comment:
i found it a little confusing that `locks` was defined below this. is that a normal pattern for scope closures?

Reply:
I think I adopted this pattern from working on JavaScript where it's a best practice to define functions above where they are called since JS hoists them anyway (and there are weird edge cases in JS where bad things happen if you don't do that).

In any case, I don't think this is a bad pattern in Python, and there are advantages to grouping together statements that are executed one after the other (and keep function definitions either before or after as scoping allows).